File: plus-one.py
# Plus One
# Time: O(n) | Space: O(1)

from typing import List

def plusOne(digits: List[int]) -> List[int]:
    # Start with carry 1 because we're adding 1 to the original number
    carry = 1
    # Starting from the end of the list, add 1
    for i in reversed(range(len(digits))):
        # If carry is 0, we can stop since rest of the list will be the same
        if carry == 0:
            return digits
        # Maffs is all around us
        curr_sum = carry + digits[i]
        # Calculate the carry
        carry = 1 if curr_sum > 9 else 0
        # Update list
        digits[i] = curr_sum % 10
    # If we have a carry left, add 1 to the beginning of the list
    if carry:
        digits = [1] + digits
    return digits

test_cases = [  [1, 2, 3],
                [4, 3, 2, 1],
                [9, 9, 9]
            ]
for case in test_cases:
    print("{0} => {1}".format(case, plusOne(case)))

File: solutions/plus-one.js
// Plus One

const test = require('./test.js');

const plusOne = digits => {
    let carry = 1;
    i = digits.length - 1;
    
    while (carry && i >= 0) {
        let digit = digits[i] + 1;
        digits[i] = digit % 10;
        carry = Math.floor(digit / 10);
        i -= 1;
    }
    
    if (carry) {
        digits.unshift(1);
    }
    
    return digits;
};

const cases = [[[1,2,3],[1,2,4]], [[4,3,2,1],[4,3,2,2]], [[9,9,9],[1,0,0,0]]];
test(cases, plusOne);
